README.md
[mldr](https://fcharte.github.io/mldr)
====

[![Travis](https://img.shields.io/travis/fcharte/mldr/master.svg)](https://travis-ci.org/fcharte/mldr/)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/mldr)](https://cran.r-project.org/package=mldr)
[![Downloads](http://cranlogs.r-pkg.org/badges/mldr)](https://cran.r-project.org/package=mldr)

Exploratory data analysis and manipulation functions for multi-label data sets along
with an interactive Shiny application to ease their use.

## Installation

Use `install.packages` to install *mldr* and its dependencies:

```R
install.packages("mldr")
```

Alternatively, you can install it via `install_github` from the
[devtools](https://github.com/hadley/devtools) package.

```R
devtools::install_github("fcharte/mldr")
```

## Building from source

Use `devtools::build` from [devtools](https://github.com/hadley/devtools)
to build the package:

```R
devtools::build(args = "--compact-vignettes=gs+qpdf")
```

## Usage and examples

This package provides a web GUI able to load, visualize and manipulate
multi-label data sets. You can launch it using the R console:

```R
library(mldr)
mldrGUI()
```

There are several functions available as well, so that you can
use *mldr* in an R script. For example, to explore some data sets:

```R
library(mldr)

# Data sets birds, emotions and genbase are
# provided within the package
print(emotions)
summary(genbase)
plot(birds)
```

*mldr* enables you to create new multi-label data sets via the
`mldr_from_dataframe` function, and export them to the standard
ARFF format using `write_arff`:

```R
library(mldr)

df <- data.frame(matrix(rnorm(1000), ncol = 10))
df$Label1 <- c(sample(c(0,1), 100, replace = TRUE))
df$Label2 <- c(sample(c(0,1), 100, replace = TRUE))
mymldr <- mldr_from_dataframe(df, labelIndices = c(11, 12), name = "testMLDR")

# Writes .arff and .xml files for a multi-label dataset
write_arff(mymldr, "my_new_mldr")
```

For more examples and detailed explanation on available functions,
please refer to the documentation.

## Citation

Please, cite *mldr* as follows:

```
@Article{charte-charte:2015,
  author       = {Francisco Charte and David Charte}, 
  title        = {Working with Multilabel Datasets in {R}: The mldr Package}, 
  journal      = {The R Journal},
  year         = 2015,
  volume       = 7,
  number       = 2,
  pages        = {149--162},
  month        = dec,
  url          = {https://journal.r-project.org/archive/2015-2/charte-charte.pdf}
}
```


NEWS.md
# Changelog

## mldr 0.4

* Reimplementation of performance evaluation metrics
* Customizable treatment of undefined values in evaluation metrics
* Support for single-quoted/double-quoted attributes in ARFF files
* Support for `stringsAsFactors`
* Support for negative `-C` parameter in MEKA header
* **API changes**: The return value of `mldr_evaluate` has the same structure but different member names.

## mldr 0.3.22

* Improve how we load available MLD's in `mldr()`
* Update `CITATION` file

## mldr 0.3.18

* Fix bug #21 when reading sparse datasets introducing zeroes instead of NAs 
* Fix #23 by ignoring case in `@attribute` tags
* Fix bug #24 in filtering function
* Improves parsing of attributes to correctly manage escaped apostrophes
* Fix calculation of SCUMBLE CV to prevent NaN values
* Optimizes some calculations (mean and CV of SCUMBLE are now 4x faster)
* Export `read.arff` function to allow reading multilabel data without calculating measures
* Add ability to load datasets from the [mldr.datasets](https://github.com/fcharte/mldr.datasets) package in `mldr()` function

## mldr 0.2.82

* Fix bug in demo code for rebuilding an mldr object
* Add global `scumble.cv` measure and `SCUMBLE.CV` measure per label
* Add new concurrence module to ease analysis of concurrence among imbalanced labels
* Display an analysis of concurrence between labels within the GUI
* Add `remedial` preprocessing algorithm
* Fix bug #18 when reading ARFF with capitalized '@relation' parameters
* A Citation file (accesible from `citation("mldr")`) has been added

## mldr 0.2.51

* Fix bug #14 when reading certain sparse datasets (e.g. Yahoo)
* Support for multiple plot types in one call
* Add `num.inputs` measure for input attributes
* Add color parameters for plotting functions

## mldr 0.2.33

* New redesigned GUI to ease usability
* Add ability to read MEKA datasets from GUI

## mldr 0.2.25

### New features

* New `mldr_evaluate()` function to assess multilabel classifier performance. Taking as input an `mldr` object and a matrix with predictions this function returns a list of metrics, including Accuracy, AUC, AveragePrecision, Coverage, FMeasure, HammingLoss, MacroAUC, MacroFMeasure, MacroPrecision, MacroRecall, MicroAUC, MicroFMeasure, MicroPrecision, MicroRecall, OneError, Precision, RankingLoss, Recall, and SubsetAccuracy
* Added more parameters to `mldr` function so labels can be identified via specific names,
indices or their count.
* Added vignette

### Fixes

* Fixed call to `chordDiagram` for newer versions of the [`circlize`](https://github.com/jokergoo/circlize) package.
* Fixed imports to avoid NOTEs on devel builds.
* Fixed parameters in calls to pROC functions.

## mldr 0.1.70

First release of *mldr*. This version includes:

* Ability to read multi-label data sets from ARFF and XML files in Mulan or MEKA format.
* Ability to write to ARFF and XML files in both Mulan and MEKA formats.
* Different ways to display data and relevant measures from data sets.
* Several plots for multi-label data visualisation.
* Functions to operate with `mldr` objects: filtering, joining and structure comparison.
* BR and LP transformations of multi-label datasets.
* Ability to create new `mldr` objects out of `data.frame`s.
* Sample datasets: `birds`, `emotions` and `genbase`.


R/mldr.R
#' @title Creates an object representing a multilabel dataset
#' @description Reads a multilabel dataset from a file and returns an \code{mldr} object
#' containing the data and additional measures. The file has to be in ARFF format.
#' The label information could be in a separate XML file (MULAN style) or in the
#' the arff header (MEKA style)
#' @param filename Name of the dataset
#' @param use_xml Specifies whether to use an
#'  associated XML file to identify the labels. Defaults to TRUE
#' @param auto_extension Specifies whether to add
#'  the '.arff' and '.xml' extensions to the filename
#'  where appropriate. Defaults to TRUE
#' @param xml_file Path to the XML file. If not
#'  provided, the filename ending in ".xml" will be
#'  assumed
#' @param label_indices Optional vector containing the indices of the attributes
#'  that should be read as labels
#' @param label_names Optional vector containing the names of the attributes
#'  that should be read as labels
#' @param label_amount Optional parameter indicating the number of labels in the
#'  dataset, which will be taken from the last attributes of the dataset
#' @param force_read_from_file Set this parameter to TRUE to always read from a local file,
#'  or set it to FALSE to look for the dataset within the `mldr.datasets` package
#' @param ... Extra parameters to be passed to 'read_arff'
#' @return An mldr object containing the multilabel dataset
#' @seealso \code{\link{mldr_from_dataframe}}, \code{\link{read.arff}}, \code{\link{summary.mldr}}
#' @examples
#'
#' library(mldr)
#'\dontrun{
#' # Read "yeast.arff" and labels from "yeast.xml"
#' mymld <- mldr("yeast")
#'
#' # Read "yeast.arff" and labels from "yeast.xml", converting categorical
#' # attributes to factors
#' mymld <- mldr("yeast", stringsAsFactors = TRUE)
#'
#' # Read "yeast-tra.arff" and labels from "yeast.xml"
#' mymld <- mldr("yeast-tra", xml_file = "yeast.xml")
#'
#' # Read "yeast.arff" specifying the amount of attributes to be used as labels
#' mymld <- mldr("yeast", label_amount = 14)
#'
#' # Read MEKA style dataset, without XML file and giving extension
#' mymld <- mldr("IMDB.arff", use_xml = FALSE, auto_extension = FALSE)
#'}
#' @import methods
#' @export
mldr <- function(filename,
                 use_xml = TRUE,
                 auto_extension = TRUE,
                 xml_file,
                 label_indices,
                 label_names,
                 label_amount,
                 force_read_from_file = !all(c(missing(xml_file), missing(label_indices), missing(label_names), missing(label_amount), use_xml, auto_extension)),
                 ...) {

  no_filename <- missing(filename)
  no_xml_file <- missing(xml_file)
  no_label_indices <- missing(label_indices)
  no_label_names <- missing(label_names)
  no_label_amount <- missing(label_amount)

  success <- FALSE

  if (!force_read_from_file) {
    if (requireNamespace("mldr.datasets", quietly = TRUE)) {
      if (exists(filename, mode = "list")) {
        ret_value <- get(filename)
        success <- TRUE
      } else if (existsFunction(filename)) {
        get(filename)()
        ret_value <- get(filename)
        success <- TRUE
      }
    }
  }

  if (success) {
    ret_value
  } else {
    if (!no_filename) {
      do.call(mldr_from_dataframe, read.arff(filename, use_xml, auto_extension, xml_file, label_indices, label_names, label_amount, ...))
    } else {
      NULL
    }
  }
}

#' @title Generates an mldr object from a data.frame and a vector with label indices
#' @description This function creates a new \code{mldr} object from the data
#' stored in a \code{data.frame}, taking as labels the columns pointed by the
#' indexes given in a vector.
#' @param dataframe The \code{data.frame} containing the dataset attributes and labels.
#' @param labelIndices Vector containing the indices of attributes acting as labels. Usually the
#' labels will be at the end (right-most columns) or the beginning (left-most columns) of the \code{data.frame}
#' @param attributes Vector with the attributes type, as returned by the  \code{attributes} member of an \code{mldr}
#' object. By default the type of the data.frame columns will be used.
#' @param name Name of the dataset. The name of the dataset given as first parameter will be used by default
#' @return An mldr object containing the multilabel dataset
#' @seealso \code{\link{mldr}}, \code{\link{summary.mldr}}
#' @examples
#'
#' library(mldr)
#'
#' df <- data.frame(matrix(rnorm(1000), ncol = 10))
#' df$Label1 <- c(sample(c(0,1), 100, replace = TRUE))
#' df$Label2 <- c(sample(c(0,1), 100, replace = TRUE))
#' mymldr <- mldr_from_dataframe(df, labelIndices = c(11, 12), name = "testMLDR")
#'
#' summary(mymldr)
#'
#' @import stats
#' @export
mldr_from_dataframe <- function(dataframe, labelIndices, attributes, name) {
  if(!is.data.frame(dataframe))
    stop(paste(substitute(dataframe), "is not a valid data.frame"))

  if(missing(labelIndices))
    stop("labelIndices parameter is compulsory")

  if(!is.numeric(labelIndices))
    stop(paste(substitute(labelIndices), "is not a numeric vector"))

  new_mldr <- list()
  new_mldr$name <- if(missing(name)) substitute(dataframe) else name
  new_mldr$dataset <- dataframe

  #new_mldr$attributesIndexes <- 1:length(dataframe)
  #new_mldr$attributesIndexes <- new_mldr$attributesIndexes[! new_mldr$attributesIndexes %in% labelIndices]
  new_mldr$attributesIndexes <- which(!1:length(dataframe) %in% labelIndices)

  if (missing(attributes)) {
    new_mldr$attributes <- sapply(new_mldr$dataset, class)
    new_mldr$attributes[labelIndices] <- "{0,1}"
    factorIndexes <- which(new_mldr$attributes == "character")

    if (length(factorIndexes > 0))
      new_mldr$attributes[factorIndexes] <- sapply(factorIndexes, function(idx)
          paste("{", paste(
              levels(as.factor(new_mldr$dataset[, idx])),
              collapse = ","
            ), "}", sep = ""
          )
        )
  } else {
    new_mldr$attributes <- attributes
  }

  new_mldr$labels <- label_measures(dataframe, labelIndices)
  new_mldr$labelsets <- if(nrow(dataframe) > 0)
                          sort(table(as.factor(do.call(paste, c(dataframe[, new_mldr$labels$index], sep = "")))))
                        else
                          array()
  new_mldr <- dataset_measures(new_mldr)
  new_mldr$measures <- measures(new_mldr)

  class(new_mldr) <- "mldr"

  new_mldr
}
